File: backend/lianyu-qq-bridge/src/main/java/com/lianyu/qqbridge/napcat/NapCatClient.java
package com.lianyu.qqbridge.napcat;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.json.JsonMapper;
import com.lianyu.qqbridge.config.QqBridgeComponent;
import com.lianyu.qqbridge.config.QqBridgeProperties;
import jakarta.annotation.PreDestroy;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.context.event.ApplicationReadyEvent;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.context.event.EventListener;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

/**
 * NapCat 正向 WebSocket 客户端：一条连接同时接收 OneBot 事件并发送 API 调用。
 * <p>
 * 职责：连接生命周期（指数退避重连 + 心跳判活看门狗）、按 {@code echo} 配对 API 请求/响应、
 * 把 {@code message} 事件以 {@link OneBotMessageEvent} 发布给 Spring（异步消费，避免阻塞 WS 收帧线程）。
 * <p>
 * 仅在 {@code lianyu.qq-bridge.enabled=true} 时装配（见 {@link QqBridgeComponent}）。
 */
@Slf4j
@QqBridgeComponent
@RequiredArgsConstructor
public class NapCatClient {

    private final QqBridgeProperties props;
    private final ApplicationEventPublisher eventPublisher;

    /** OneBot JSON 为 snake_case；用独立 mapper，不影响全局 ObjectMapper。 */
    private final ObjectMapper mapper = JsonMapper.builder()
            .findAndAddModules()
            .propertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)
            .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
            .build();

    private final HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(15))
            .build();

    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
        Thread t = new Thread(r, "qq-bridge-scheduler");
        t.setDaemon(true);
        return t;
    });

    private final AtomicReference<WebSocket> socket = new AtomicReference<>();
    private final Map<String, CompletableFuture<JsonNode>> pending = new ConcurrentHashMap<>();
    private final AtomicBoolean running = new AtomicBoolean(false);
    private final AtomicBoolean reconnecting = new AtomicBoolean(false);
    private final AtomicInteger attempt = new AtomicInteger(0);
    private volatile long selfId = 0L;
    private volatile Instant lastFrameAt = Instant.now();
    private ScheduledFuture<?> watchdog;

    @EventListener(ApplicationReadyEvent.class)
    public void start() {
        if (!running.compareAndSet(false, true)) {
            return;
        }
        log.info("QQ bridge starting, target={}", maskUrl(appendToken(props.getNapcat().getWsUrl(), props.getNapcat().getAccessToken())));
        connect();
        watchdog = scheduler.scheduleAtFixedRate(this::watchdog, 30, 30, TimeUnit.SECONDS);
    }

    @PreDestroy
    public void stop() {
        running.set(false);
        if (watchdog != null) {
            watchdog.cancel(false);
        }
        WebSocket ws = socket.getAndSet(null);
        if (ws != null) {
            try {
                ws.sendClose(WebSocket.NORMAL_CLOSURE, "shutdown");
            } catch (Exception ignored) {
                // 忽略关闭异常
            }
        }
        failAllPending("bridge shutting down");
        scheduler.shutdownNow();
        log.info("QQ bridge stopped");
    }

    private void connect() {
        String fullUrl = appendToken(props.getNapcat().getWsUrl(), props.getNapcat().getAccessToken());
        try {
            log.info("QQ bridge connecting to {}", maskUrl(fullUrl));
            httpClient.newWebSocketBuilder()
                    .connectTimeout(Duration.ofSeconds(Math.max(1, props.getNapcat().getConnectTimeoutSeconds())))
                    .buildAsync(URI.create(fullUrl), new FrameListener())
                    .thenAccept(ws -> {
                        socket.set(ws);
                        attempt.set(0);
                        lastFrameAt = Instant.now();
                        log.info("QQ bridge WebSocket connected");
                        onConnect();
                    })
                    .exceptionally(t -> {
                        log.warn("QQ bridge connect failed: {}", String.valueOf(t));
                        scheduleReconnect();
                        return null;
                    });
        } catch (Exception e) {
            log.warn("QQ bridge connect error: {}", String.valueOf(e));
            scheduleReconnect();
        }
    }

    private void onConnect() {
        // 探测自身 QQ 号，用于群聊识别 @
        sendApi("get_login_info", Map.of())
                .whenComplete((data, err) -> {
                    if (err != null) {
                        log.warn("QQ bridge get_login_info failed: {}", String.valueOf(err));
                        return;
                    }
                    try {
                        long uid = data.path("user_id").asLong(0L);
                        if (uid > 0) {
                            selfId = uid;
                            log.info("QQ bridge logged in as {}", uid);
                        }
                    } catch (Exception e) {
                        log.warn("QQ bridge get_login_info parse failed: {}", String.valueOf(e));
                    }
                });
    }

    private void onDisconnect() {
        socket.set(null);
        failAllPending("connection closed");
        scheduleReconnect();
    }

    private void scheduleReconnect() {
        if (!running.get()) {
            return;
        }
        if (!reconnecting.compareAndSet(false, true)) {
            return;
        }
        int n = attempt.incrementAndGet();
        long base = Math.min(30_000L, 1_000L * (1L << Math.min(n, 5)));
        long delay = base + ThreadLocalRandom.current().nextLong(0, 1_000);
        scheduler.schedule(() -> {
            reconnecting.set(false);
            connect();
        }, delay, TimeUnit.MILLISECONDS);
    }

    private void watchdog() {
        try {
            if (!running.get() || socket.get() == null) {
                return;
            }
            long since = Duration.between(lastFrameAt, Instant.now()).getSeconds();
            int timeout = Math.max(30, props.getNapcat().getHeartbeatTimeoutSeconds());
            if (since > timeout) {
                log.warn("QQ bridge no frame for {}s (>{}s), force reconnect", since, timeout);
                WebSocket ws = socket.getAndSet(null);
                if (ws != null) {
                    try {
                        ws.sendClose(WebSocket.NORMAL_CLOSURE, "watchdog");
                    } catch (Exception ignored) {
                        // 忽略关闭异常
                    }
                }
                failAllPending("watchdog reconnect");
                scheduleReconnect();
            }
        } catch (Exception e) {
            log.warn("QQ bridge watchdog error: {}", String.valueOf(e));
        }
    }

    private void handleFrame(String json) {
        lastFrameAt = Instant.now();
        try {
            JsonNode root = mapper.readTree(json);
            JsonNode echoNode = root.get("echo");
            if (echoNode != null && !echoNode.isNull()) {
                String echo = echoNode.asText();
                CompletableFuture<JsonNode> future = pending.remove(echo);
                if (future != null) {
                    JsonNode data = root.get("data");
                    if (data == null || data.isNull()) {
                        data = mapper.getNodeFactory().objectNode();
                    }
                    int retcode = root.path("retcode").asInt(0);
                    if (retcode != 0) {
                        future.completeExceptionally(new IllegalStateException(
                                "OneBot API retcode=" + retcode + " status=" + root.path("status").asText()));
                    } else {
                        future.complete(data);
                    }
                } else {
                    log.debug("QQ bridge received API response with unknown echo: {}", echo);
                }
                return;
            }
            String postType = root.path("post_type").asText("");
            if ("message".equals(postType)) {
                OneBotModels.MessageEvent ev = mapper.treeToValue(root, OneBotModels.MessageEvent.class);
                eventPublisher.publishEvent(new OneBotMessageEvent(this, ev));
            } else if ("meta_event".equals(postType)) {
                // heartbeat / lifecycle：仅作活性判据（lastFrameAt 已更新），无需处理
            } else {
                log.debug("QQ bridge ignored OneBot event post_type={}", postType);
            }
        } catch (Exception e) {
            log.warn("QQ bridge failed to handle frame: {} | frame={}", String.valueOf(e), abbreviate(json, 500));
        }
    }

    /**
     * 发送一次 OneBot API 调用，返回其 {@code data} 字段的 Future（按 {@code echo} 配对，30s 超时）。
     */
    public CompletableFuture<JsonNode> sendApi(String action, Map<String, Object> params) {
        CompletableFuture<JsonNode> future = new CompletableFuture<>();
        WebSocket ws = socket.get();
        if (ws == null) {
            future.completeExceptionally(new IllegalStateException("QQ bridge not connected"));
            return future;
        }
        String echo = UUID.randomUUID().toString();
        pending.put(echo, future);
        String payload;
        try {
            payload = mapper.writeValueAsString(Map.of(
                    "action", action,
                    "params", params == null ? Map.of() : params,
                    "echo", echo));
        } catch (Exception e) {
            pending.remove(echo);
            future.completeExceptionally(e);
            return future;
        }
        try {
            ws.sendText(payload, true);
        } catch (Exception e) {
            pending.remove(echo);
            future.completeExceptionally(e);
            return future;
        }
        scheduler.schedule(() -> {
            CompletableFuture<JsonNode> f = pending.remove(echo);
            if (f != null) {
                f.completeExceptionally(new TimeoutException("QQ API timeout: " + action));
            }
        }, 30, TimeUnit.SECONDS);
        return future;
    }

    /** 发送私聊文本（fire-and-forget，失败仅告警）。 */
    public void sendPrivateMsg(long userId, String text) {
        sendApi("send_private_msg", Map.of(
                        "user_id", userId,
                        "message", List.of(Map.of("type", "text", "data", Map.of("text", text)))))
                .whenComplete((d, e) -> {
                    if (e != null) {
                        log.warn("QQ bridge send_private_msg failed for {}: {}", userId, String.valueOf(e));
                    }
                });
    }

    /** 发送群聊文本（fire-and-forget，失败仅告警）。 */
    public void sendGroupMsg(long groupId, String text) {
        sendApi("send_group_msg", Map.of(
                        "group_id", groupId,
                        "message", List.of(Map.of("type", "text", "data", Map.of("text", text)))))
                .whenComplete((d, e) -> {
                    if (e != null) {
                        log.warn("QQ bridge send_group_msg failed for {}: {}", groupId, String.valueOf(e));
                    }
                });
    }

    /** 当前登录的机器人 QQ 号（get_login_info 探测结果，未探测到为 0）。 */
    public long getSelfId() {
        return selfId;
    }

    private void failAllPending(String reason) {
        if (pending.isEmpty()) {
            return;
        }
        pending.values().forEach(f -> f.completeExceptionally(new IllegalStateException(reason)));
        pending.clear();
    }

    private static String appendToken(String url, String token) {
        if (url == null || url.isBlank()) {
            return url;
        }
        if (token == null || token.isBlank()) {
            return url;
        }
        String sep = url.contains("?") ? "&" : "?";
        // token 进入 query 须 URL 编码：未编码的 & # = 空格等会注入或截断 query 参数，
        // 既可能让 NapCat 解析出错的 access_token，也可能拼出非预期键值
        return url + sep + "access_token=" + URLEncoder.encode(token, StandardCharsets.UTF_8);
    }

    private static String maskUrl(String url) {
        return url == null ? null : url.replaceAll("access_token=[^&]*", "access_token=***");
    }

    private static String abbreviate(String s, int max) {
        if (s == null) {
            return null;
        }
        return s.length() <= max ? s : s.substring(0, max) + "...(" + s.length() + ")";
    }

    /**
     * OneBot WS 帧监听器：文本帧可能分多片到达，按 {@code last} 拼接后整体处理。
     * 所有异常路径都触发重连，绝不让异常冒泡到 JDK WS 框架导致静默断连。
     */
    private final class FrameListener implements WebSocket.Listener {
        private final StringBuilder buf = new StringBuilder();

        @Override
        public void onOpen(WebSocket webSocket) {
            webSocket.request(1);
        }

        @Override
        public java.util.concurrent.CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
            buf.append(data);
            if (last) {
                String frame = buf.toString();
                buf.setLength(0);
                handleFrame(frame);
            }
            webSocket.request(1);
            return null;
        }

        @Override
        public java.util.concurrent.CompletionStage<?> onClose(WebSocket webSocket, int statusCode, String reason) {
            log.info("QQ bridge WebSocket closed: {} {}", statusCode, reason);
            onDisconnect();
            return null;
        }

        @Override
        public void onError(WebSocket webSocket, Throwable error) {
            log.warn("QQ bridge WebSocket error: {}", String.valueOf(error));
            onDisconnect();
        }

        @Override
        public java.util.concurrent.CompletionStage<?> onPing(WebSocket webSocket, ByteBuffer message) {
            try {
                webSocket.sendPong(message);
            } catch (Exception ignored) {
                // 忽略回 pong 异常
            }
            return null;
        }
    }

}

File: frontend/electron/qqBridge/napCatClient.js
/**
 * NapCat 正向 WebSocket 客户端（OneBot 11）。
 *
 * 与后端 lianyu-qq-bridge 的 NapCatClient（Java）同构：
 * - echo 配对 API 请求/响应（sendApi 返回 Promise<data>）
 * - 收到 post_type==="message" 的事件经 onMessage 回调上抛
 * - 断线指数退避重连（base * 2^min(n,5)，封顶 reconnectMaxMs）
 *
 * 仅用于主进程直连本地 NapCat（ws://127.0.0.1:...），不经过 Electron net/证书 pin；
 * 云端 API 调用另走 apiProxy.performApiRequest。
 */
import WebSocket from 'ws'

const API_TIMEOUT_MS = 15000

export function createNapCatClient({
  wsUrl,
  accessToken = '',
  connectTimeoutMs = 10000,
  reconnectBaseMs = 600,
  reconnectMaxMs = 6000,
  onMessage,
  onStatus,
} = {}) {
  let ws = null
  let selfId = ''
  let currentState = 'stopped'
  let echoSeq = 0
  const pending = new Map() // echo -> { resolve, reject, timer }
  let reconnectTimer = null
  let connectTimer = null
  let selfIdProbeTimer = null
  let probeEpoch = 0 // 递增计数器：kick 事件递增，使 in-flight 的 get_login_info 响应作废
  let reconnectAttempts = 0
  let stopped = true
  let connecting = false

  function emitStatus(state, extra = {}) {
    currentState = state
    try {
      onStatus?.({ state, selfId, ...extra })
    } catch {
      /* 回调异常不影响连接 */
    }
  }

  function clearReconnectTimer() {
    if (reconnectTimer) {
      clearTimeout(reconnectTimer)
      reconnectTimer = null
    }
  }

  function clearConnectTimer() {
    if (connectTimer) {
      clearTimeout(connectTimer)
      connectTimer = null
    }
  }

  // 登录态探测 + 在线 watchdog：
  // - 未登录时每 5s 快速补探，拿到 selfId 即升 ready；
  // - 已登录后切到 30s 慢速 watchdog，验证 QQ 是否仍在线；
  // - watchdog 发现 get_login_info 返回空（QQ 被踢下线）→ 清 selfId、降级 connected、重启快速补探。
  // 断线/停止时一并清掉定时器。
  function clearSelfIdProbe() {
    if (selfIdProbeTimer) {
      clearInterval(selfIdProbeTimer)
      selfIdProbeTimer = null
    }
  }

  function probeSelfId() {
    if (stopped || !ws || ws.readyState !== WebSocket.OPEN) return
    const epoch = probeEpoch
    sendApi('get_login_info')
      .then((data) => {
        if (epoch !== probeEpoch) return // 过时响应（kick 事件已递增 epoch），丢弃
        const id = data && data.user_id != null ? String(data.user_id) : ''
        if (id) {
          if (!selfId) {
            // 首次拿到 selfId：升级到 ready，切慢速 watchdog
            selfId = id
            emitStatus('ready', { selfId })
            clearSelfIdProbe()
            selfIdProbeTimer = setInterval(probeSelfId, 30000)
          }
          // selfId 已存在且仍在线：无需变化
        } else if (selfId) {
          // QQ 下线了（之前有 selfId，现在 get_login_info 返回空）
          selfId = ''
          emitStatus('connected', { kicked: true })
          clearSelfIdProbe()
          selfIdProbeTimer = setInterval(probeSelfId, 5000)
        }
      })
      .catch(() => {
        /* 未登录或暂不可用，下个周期再探 */
      })
  }

  function startSelfIdProbe() {
    clearSelfIdProbe()
    selfIdProbeTimer = setInterval(probeSelfId, 5000)
    probeSelfId() // 立即探一次，缩短感知延迟
  }

  function failPending(error) {
    for (const { reject, timer } of pending.values()) {
      clearTimeout(timer)
      reject(error)
    }
    pending.clear()
  }

  function buildWsUrl(baseUrl, token) {
    if (!token) return baseUrl
    try {
      const u = new URL(baseUrl)
      if (!u.searchParams.has('access_token')) {
        u.searchParams.set('access_token', token)
      }
      return u.toString()
    } catch {
      const sep = baseUrl.includes('?') ? '&' : '?'
      return `${baseUrl}${sep}access_token=${encodeURIComponent(token)}`
    }
  }

  function start() {
    stopped = false
    connect()
  }

  function stop() {
    stopped = true
    clearReconnectTimer()
    clearConnectTimer()
    clearSelfIdProbe()
    failPending(new Error('napcat client stopped'))
    if (ws) {
      try {
        ws.removeAllListeners()
        ws.close()
      } catch {
        /* ignore */
      }
      ws = null
    }
    selfId = ''
    emitStatus('stopped')
  }

  function connect() {
    if (stopped || connecting) return
    if (ws && (ws.readyState === WebSocket.OPEN || ws.readyState === WebSocket.CONNECTING)) return
    connecting = true
    clearConnectTimer()
    emitStatus('connecting')

    let socket
    try {
      socket = new WebSocket(buildWsUrl(wsUrl, accessToken))
    } catch (e) {
      connecting = false
      emitStatus('error', { error: e?.message || String(e) })
      scheduleReconnect()
      return
    }
    ws = socket

    // 自管握手超时：到期仍未 OPEN 则强断，触发 onClose 走重连
    connectTimer = setTimeout(() => {
      if (socket.readyState !== WebSocket.OPEN) {
        try {
          socket.terminate()
        } catch {
          /* ignore */
        }
      }
    }, Math.max(1000, connectTimeoutMs))

    socket.on('open', onOpen)
    socket.on('message', onRawMessage)
    socket.on('close', onClose)
    socket.on('error', onError)
  }

  function onOpen() {
    connecting = false
    clearConnectTimer()
    reconnectAttempts = 0
    emitStatus('connected')
    // 拉取自身 QQ 号，确认登录态；失败仍保持连接（消息照收）
    sendApi('get_login_info')
      .then((data) => {
        selfId = data && data.user_id != null ? String(data.user_id) : ''
        if (selfId) {
          emitStatus('ready', { selfId })
          // 已登录：启动 30s 慢速 watchdog，监测 QQ 是否被踢下线
          clearSelfIdProbe()
          selfIdProbeTimer = setInterval(probeSelfId, 30000)
        } else {
          // 未登录：开启补探，等用户登录 QQ 后自动升 ready（修复登录后状态长时间不变绿）
          startSelfIdProbe()
        }
      })
      .catch((e) => {
        console.warn('[qqBridge] get_login_info failed:', e?.message || e)
        startSelfIdProbe()
      })
  }

  function onRawMessage(raw) {
    let payload
    try {
      payload = JSON.parse(raw.toString('utf-8'))
    } catch {
      return
    }
    // API 响应：按 echo 配对兑现
    const echo = payload?.echo
    if (echo != null && pending.has(echo)) {
      const entry = pending.get(echo)
      pending.delete(echo)
      clearTimeout(entry.timer)
      if (payload.retcode != null && payload.retcode !== 0) {
        entry.reject(
          new Error(`napcat api error: retcode=${payload.retcode} ${payload.msg || payload.wording || ''}`.trim()),
        )
      } else {
        entry.resolve(payload.data)
      }
      return
    }
    // 事件：仅上抛消息事件
    if (payload.post_type === 'message') {
      try {
        onMessage?.(payload)
      } catch (e) {
        console.warn('[qqBridge] onMessage handler error:', e?.message || e)
      }
      return
    }
    // meta_event / notice：检测 QQ 下线（被踢/主动下线/网络断开）
    // NapCat 在 QQ 被踢时会发 lifecycle disable 或自定义 notice，get_login_info 可能仍返回缓存 user_id，
    // 故必须靠事件检测，不能只靠轮询。
    // 注意：sub_type 里的 'kick' 不能裸匹配——group_decrease 群成员被踢的 sub_type 也是 'kick'，
    // 那是群事件不是机器人掉线。只匹配 notice_type 含 offline、或 sub_type 为 disable/offline。
    if (payload.post_type === 'meta_event' || payload.post_type === 'notice') {
      const sub = payload.sub_type || ''
      const ntype = payload.notice_type || ''
      const isOffline = (
        (payload.meta_event_type === 'lifecycle' && sub === 'disable') ||
        /offline/i.test(ntype) ||
        /offline|disable/i.test(sub)
      )
      if (isOffline && selfId) {
        console.warn('[napcat] QQ offline event:', payload.meta_event_type || ntype, sub)
        selfId = ''
        probeEpoch++ // 使可能 in-flight 的 get_login_info 响应作废，防止撤销掉线检测
        clearSelfIdProbe()
        emitStatus('connected', { kicked: true })
        selfIdProbeTimer = setInterval(probeSelfId, 5000)
      }
      return
    }
  }

  function onClose(code, reasonBuf) {
    connecting = false
    clearConnectTimer()
    clearSelfIdProbe()
    failPending(new Error('napcat ws closed'))
    const reason = reasonBuf ? reasonBuf.toString('utf-8') : ''
    if (!stopped) {
      emitStatus('disconnected', { code, reason })
      scheduleReconnect()
    }
  }

  function onError(err) {
    // 'close' 会随后触发并负责重连；此处仅记录，避免重复调度
    console.warn('[qqBridge] ws error:', err?.message || err)
    connecting = false
    clearConnectTimer()
  }

  function scheduleReconnect() {
    if (stopped) return
    clearReconnectTimer()
    const exp = Math.min(reconnectBaseMs * (1 << Math.min(reconnectAttempts, 5)), reconnectMaxMs)
    reconnectAttempts += 1
    emitStatus('reconnecting', { attempt: reconnectAttempts, delayMs: exp })
    reconnectTimer = setTimeout(() => {
      reconnectTimer = null
      connect()
    }, exp)
  }

  function sendApi(action, params = {}, timeoutMs = API_TIMEOUT_MS) {
    return new Promise((resolve, reject) => {
      if (!ws || ws.readyState !== WebSocket.OPEN) {
        reject(new Error(`napcat ws not open: ${action}`))
        return
      }
      const echo = String(++echoSeq)
      const timer = setTimeout(() => {
        if (pending.delete(echo)) reject(new Error(`napcat api timeout: ${action}`))
      }, timeoutMs)
      pending.set(echo, { resolve, reject, timer })
      try {
        ws.send(JSON.stringify({ action, params, echo }))
      } catch (e) {
        pending.delete(echo)
        clearTimeout(timer)
        reject(e)
      }
    })
  }

  function sendPrivateMsg(userId, text) {
    return sendApi('send_private_msg', {
      user_id: Number(userId),
      message: [{ type: 'text', data: { text } }],
    })
  }

  function sendGroupMsg(groupId, text) {
    return sendApi('send_group_msg', {
      group_id: Number(groupId),
      message: [{ type: 'text', data: { text } }],
    })
  }

  // 发送图片（OneBot 11 image 段）：NapCat 支持 base64:// 内联，免去公网回拉。
  // base64 须不含 data:image/... 前缀，仅裸 base64。
  function sendPrivateMsgImage(userId, base64) {
    return sendApi('send_private_msg', {
      user_id: Number(userId),
      message: [{ type: 'image', data: { file: `base64://${base64}` } }],
    })
  }

  function sendGroupMsgImage(groupId, base64) {
    return sendApi('send_group_msg', {
      group_id: Number(groupId),
      message: [{ type: 'image', data: { file: `base64://${base64}` } }],
    })
  }

  function getSelfId() {
    return selfId
  }

  function getStatus() {
    return { state: currentState, selfId }
  }

  return { start, stop, sendApi, sendPrivateMsg, sendGroupMsg, sendPrivateMsgImage, sendGroupMsgImage, getSelfId, getStatus }
}
